In [7]:
#importing pandas
import pandas as pd

In [17]:
#getting out dataset
iriscsv=pd.read_csv("https://raw.githubusercontent.com/mpourhoma/CS4661/master/iris.csv")

In [20]:
iriscsv

sepal_length  sepal_width  petal_length  petal_width    species
0             5.1          3.5           1.4          0.2     setosa
1             4.9          3.0           1.4          0.2     setosa
2             4.7          3.2           1.3          0.2     setosa
3             4.6          3.1           1.5          0.2     setosa
4             5.0          3.6           1.4          0.2     setosa
5             5.4          3.9           1.7          0.4     setosa
6             4.6          3.4           1.4          0.3     setosa
7             5.0          3.4           1.5          0.2     setosa
8             4.4          2.9           1.4          0.2     setosa
9             4.9          3.1           1.5          0.1     setosa
10            5.4          3.7           1.5          0.2     setosa
11            4.8          3.4           1.6          0.2     setosa
12            4.8          3.0           1.4          0.1     setosa
13            4.3          3.0           1.1          0.1     setosa
14            5.8          4.0           1.2          0.2     setosa
15            5.7          4.4           1.5          0.4     setosa
16            5.4          3.9           1.3          0.4     setosa
17            5.1          3.5           1.4          0.3     setosa
18            5.7          3.8           1.7          0.3     setosa
19            5.1          3.8           1.5          0.3     setosa
20            5.4          3.4           1.7          0.2     setosa
21            5.1          3.7           1.5          0.4     setosa
22            4.6          3.6           1.0          0.2     setosa
23            5.1          3.3           1.7          0.5     setosa
24            4.8          3.4           1.9          0.2     setosa
25            5.0          3.0           1.6          0.2     setosa
26            5.0          3.4           1.6          0.4     setosa
27            5.2          3.5           1.5          0.2     setosa
28            5.2          3.4           1.4          0.2     setosa
29            4.7          3.2           1.6          0.2     setosa
..            ...          ...           ...          ...        ...
120           6.9          3.2           5.7          2.3  virginica
121           5.6          2.8           4.9          2.0  virginica
122           7.7          2.8           6.7          2.0  virginica
123           6.3          2.7           4.9          1.8  virginica
124           6.7          3.3           5.7          2.1  virginica
125           7.2          3.2           6.0          1.8  virginica
126           6.2          2.8           4.8          1.8  virginica
127           6.1          3.0           4.9          1.8  virginica
128           6.4          2.8           5.6          2.1  virginica
129           7.2          3.0           5.8          1.6  virginica
130           7.4          2.8           6.1          1.9  virginica
131           7.9          3.8           6.4          2.0  virginica
132           6.4          2.8           5.6          2.2  virginica
133           6.3          2.8           5.1          1.5  virginica
134           6.1          2.6           5.6          1.4  virginica
135           7.7          3.0           6.1          2.3  virginica
136           6.3          3.4           5.6          2.4  virginica
137           6.4          3.1           5.5          1.8  virginica
138           6.0          3.0           4.8          1.8  virginica
139           6.9          3.1           5.4          2.1  virginica
140           6.7          3.1           5.6          2.4  virginica
141           6.9          3.1           5.1          2.3  virginica
142           5.8          2.7           5.1          1.9  virginica
143           6.8          3.2           5.9          2.3  virginica
144           6.7          3.3           5.7          2.5  virginica
145           6.7          3.0           5.2          2.3  virginica
146

In [21]:
feat= ['sepal_length','sepal_width','petal_length','petal_width']
a= iriscsv[feat]  
i=iriscsv['species']

In [22]:
from sklearn.model_selection import train_test_split

a_train,a_test,i_train,i_test = train_test_split (a,i,test_size=0.4,random_state=6)


In [23]:
from sklearn.neighbors import KNeighborsClassifier

In [24]:
k=3
knn_hw2_c= KNeighborsClassifier(n_neighbors=k)

In [25]:
knn_hw2_c.fit(a_train,i_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=3, p=2,
           weights='uniform')

In [26]:
i_predict =knn_hw2_c.predict(a_test)

print(i_predict)

['setosa' 'virginica' 'setosa' 'setosa' 'virginica' 'versicolor'
 'virginica' 'setosa' 'virginica' 'versicolor' 'virginica' 'versicolor'
 'virginica' 'virginica' 'versicolor' 'versicolor' 'virginica'
 'versicolor' 'versicolor' 'setosa' 'setosa' 'virginica' 'setosa' 'setosa'
 'versicolor' 'versicolor' 'versicolor' 'virginica' 'setosa' 'versicolor'
 'setosa' 'versicolor' 'setosa' 'setosa' 'versicolor' 'virginica'
 'versicolor' 'virginica' 'versicolor' 'setosa' 'setosa' 'virginica'
 'versicolor' 'versicolor' 'setosa' 'setosa' 'versicolor' 'setosa'
 'setosa' 'versicolor' 'virginica' 'virginica' 'virginica' 'setosa'
 'virginica' 'setosa' 'setosa' 'setosa' 'versicolor' 'virginica']


In [27]:
# Function "accuracy_score" from "sklearn.metrics" will perform element-to-element comparision and returns the 
# percent of correct predictions:

from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test,y_predict)

print(accuracy)

NameError: name 'y_test' is not defined

### (D) Repeat	part (c)		for	K=1,	K=5,	K=7,	K=15,	K=28,	K=60	(you	can	simply	write	a	“for	loop”	on	a	list	with	these	values	to	repeat	the	process for	you,	and	save	the	final	accuracy	results	 in	a	list	(array)!).	Does	the	accuracy	always	get	better	by	increasing	K?	


In [28]:
# Defined list of K with name Knn_list and empty list to store each individuals accuracy

knn_list=[1,5,7,11,15,27,59]

knn_result_d={}

# using for loop to perform KNeighborsClassifier on different value of K and appending accuracy in knn_result_d list

for k in knn_list:
    
    knn_hw2_d= KNeighborsClassifier(n_neighbors=k)
    
    knn_hw2_d.fit(a_train,i_train)
    
    i_predict_d =knn_hw2_d.predict(a_test)
    
    accuracy =accuracy_score(i_test,i_predict_d)
    
    knn_result_d[k]=accuracy
        
    
print(knn_result_d)

{1: 0.95, 5: 0.9833333333333333, 7: 0.9666666666666667, 11: 0.9666666666666667, 15: 0.9333333333333333, 27: 0.9166666666666666, 59: 0.8166666666666667}


 ## Does the accuracy always get better by increasing K?
 
 ## Answer: No

### (E) Now,	suppose	that	we	would	like	to	make	prediction	based	on	only	ONE	single	feature.	To	find	the	best	single	feature,	we	have	to	try	every	feature	individually.	In	other	word,	we	 have	 to	 repeat	 part (c)	 4	 times	 (each	 time	 using	 only	 one	 of	 the	 4	 features),	 and	compute	 the	accuracy	each	 time.	Then,	check	 the	accuracies.	Which	individual	 feature	provide	 the	 best	 accuracy	 (the	 best	 feature)?	 Which	 one	 is	 the	 second	best	feature?	(Note:	you	have	to	train,	test,	and	evaluate your	model	4	times,	each	time	on	a	datasetincluding	only	one	of	the	features.	You	may	want	to	write	a	“for	loop”	to	generate	the	datasets,	 train,	 test,	and	compute	 the	accuracy	each	 time,	and	save	 the	 final	accuracy	results	in	a	list).

In [37]:
# we perform the KNeighborsClassifier on individual features and store each one's accuracy in knn_result_dict dictonary 

k=3

knn_hw2_e= KNeighborsClassifier(n_neighbors=k)

knn_result_dict={}

for f in feat:
    
        a_feature=iriscsv[f]

        a_ftrain,a_ftest,i_ftrain,i_ftest = train_test_split (a_feature.values.reshape(len(a_feature),1),i,test_size=0.4,random_state=1)

        knn_hw2_e.fit(a_ftrain,i_ftrain)

        i_fpredict=knn_hw2_e.predict(a_ftest)

        accuracy = accuracy_score(i_ftest,i_fpredict)

        knn_result_dict[f]=accuracy
        
    
print(knn_result_dict)
      

    

{'sepal_length': 0.6333333333333333, 'sepal_width': 0.5333333333333333, 'petal_length': 0.9666666666666667, 'petal_width': 0.9333333333333333}


## Which Individual features provides the best accuracy(the best feature)?
### Answer: petal_length with 0.966666666667 accuracy
## Which one is the second best features ?
### Answer: petal_width with 0.93333333333333335  accuracy

### (F)  Now,	we	want	to	repeat	part	(e),	this	time	using	two	features.	you	have	to	train,	test,	and	 evaluate	 your	 model	 for	 6	 different	 cases:	 using	 (1st and	 2nd features),	 (1st and	 3rd features),	(1st and	4th		features),	(2nd		and	3rd		features), (2nd and	4th features), (3rd and	4th	features)!																															Which “feature	pair” provide	the	best	accuracy?

In [39]:

# we perform the KNeighborsClassifier on different feature pairs and store each one's accuracy in knn_result 

k=3

knn_hw2_e= KNeighborsClassifier(n_neighbors=k)

features =[['sepal_length','sepal_width'],['sepal_length','petal_length'],['sepal_length','petal_width'],['sepal_width','petal_length'],['sepal_width','petal_width'],['petal_length','petal_width']]

knn_result={}

for f in features:
    
        a_feature=iriscsv[f]

        a_ftrain,a_ftest,i_ftrain,i_ftest = train_test_split (a_feature,i,test_size=0.4,random_state=1)

        knn_hw2_e.fit(a_ftrain,i_ftrain)

        i_fpredict=knn_hw2_e.predict(a_ftest)

        accuracy = accuracy_score(i_ftest,i_fpredict)

        knn_result[','.join(f)]=accuracy

print(knn_result)



for k in knn_result.keys():
        
        print(k,":",knn_result[k])
    

{'sepal_length,sepal_width': 0.65, 'sepal_length,petal_length': 0.9333333333333333, 'sepal_length,petal_width': 0.9833333333333333, 'sepal_width,petal_length': 0.95, 'sepal_width,petal_width': 0.9666666666666667, 'petal_length,petal_width': 0.9666666666666667}
sepal_length,sepal_width : 0.65
sepal_length,petal_length : 0.9333333333333333
sepal_length,petal_width : 0.9833333333333333
sepal_width,petal_length : 0.95
sepal_width,petal_width : 0.9666666666666667
petal_length,petal_width : 0.9666666666666667


# Which “feature	pair” provide	the	best	accuracy?

## Ans: feature pair ('sepal_length','petal_width') with 0.983333333333 accuracy

### (G) BigQuestion: Does the	“best	feature	pair” from	part	(f) contain of both “first best	feature”and	“second	best	feature” from	part	(e)?	 In	other	word,	can	we	conclude	that	the	“best	two	features”	for	classification	are	the	“first	best	individual	feature”	and	the	“second	best	individual	feature”	together?


#### Ans: No, "best features pair" (sepal_length, petal_width) with accuracy 0.983333333333 does not contain the  first best feature "petal_length" and second best feature "petal_width", So best feature pair is more accurate than first and second best  individual features, we can say that "best two features" for classification are not “first	best	individual	feature”	and	the	“second	best	individual	feature”	together.

### (H) Optional	Question:	Justify	your	answer	to	part	(g)!		If	yes,	why?		If	no,	why	not?

#### Ans. When we take one feature at one time then we consider only that feature an it's distance from testing data. while we take pair of features then we have to consider both feature and it's distance from testing sample .so, everytime it is not necessary that first and second individual feature together give a best accuracy.